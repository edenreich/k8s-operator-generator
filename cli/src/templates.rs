use askama::Template;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

mod filters {
    pub fn dashcase<T: std::fmt::Display>(s: T) -> ::askama::Result<String> {
        Ok(s.to_string()
            .split_whitespace()
            .collect::<Vec<&str>>()
            .join("-")
            .to_lowercase())
    }
}

// Common Identifiers and Structs
pub struct RoleTemplateIdentifiers {
    pub api_group: String,
    pub resources: Vec<String>,
}

pub struct ClusterRoleTemplateIdentifiers {
    pub api_group: String,
    pub resources: Vec<String>,
}

#[derive(Serialize, Deserialize)]
pub struct Resource {
    pub api_group: String,
    pub api_version: String,
    pub kind: String,
    pub metadata: Metadata,
    pub spec: String,
}

#[derive(Serialize, Deserialize)]
pub struct Metadata {
    pub name: String,
}

pub struct Field {
    pub pub_name: String,
    pub field_type: String,
}

pub struct ControllerAttributes {
    pub operation_id: String,
    pub http_method: String,
    pub action_summary: String,
}

// Tests Templates
pub mod tests {
    use super::*;

    #[derive(Template)]
    #[template(path = "tests/main.rs.jinja")]
    pub struct Main {}

    #[derive(Template)]
    #[template(path = "tests/utils_client.rs.jinja")]
    pub struct UtilsClient {}

    #[derive(Template)]
    #[template(path = "tests/utils_operator.rs.jinja")]
    pub struct UtilsOperator {}

    #[derive(Template)]
    #[template(path = "tests/utils_cluster.rs.jinja")]
    pub struct UtilsCluster {}
}

// Devcontainer Templates
pub mod devcontainer {
    use super::*;

    #[derive(Template)]
    #[template(path = ".devcontainer/devcontainer.json.jinja")]
    pub struct Json {}

    #[derive(Template)]
    #[template(path = ".devcontainer/dockerfile.jinja")]
    pub struct Dockerfile {}

    #[derive(Template)]
    #[template(path = ".devcontainer/setup-git.sh.jinja")]
    pub struct SetupGit {}

    #[derive(Template)]
    #[template(path = ".devcontainer/launch.json.jinja")]
    pub struct LaunchJsonExample {}

    #[derive(Template)]
    #[template(path = ".devcontainer/zshrc.jinja")]
    pub struct Zshrc {}
}

// Manifests Templates
pub mod manifests {
    use super::*;

    pub mod rbac {
        use super::*;

        #[derive(Template)]
        #[template(path = "manifests/rbac_role.yaml.jinja")]
        pub struct Role {
            pub identifiers: RoleTemplateIdentifiers,
        }

        #[derive(Template)]
        #[template(path = "manifests/rbac_cluster_role.yaml.jinja")]
        pub struct ClusterRole {
            pub identifiers: ClusterRoleTemplateIdentifiers,
        }

        #[derive(Template)]
        #[template(path = "manifests/rbac_service_account.yaml.jinja")]
        pub struct ServiceAccount {}

        #[derive(Template)]
        #[template(path = "manifests/rbac_role_binding.yaml.jinja")]
        pub struct RoleBinding {}

        #[derive(Template)]
        #[template(path = "manifests/rbac_cluster_role_binding.yaml.jinja")]
        pub struct ClusterRoleBinding {}
    }

    pub mod operator {
        use super::*;
        #[derive(Template)]
        #[template(path = "manifests/operator_deployment.yaml.jinja")]
        pub struct Deployment {
            pub secret_name: String,
        }

        #[derive(Template)]
        #[template(path = "manifests/operator_secret.yaml.jinja")]
        pub struct Secret {
            pub secret_name: String,
        }
    }

    pub mod examples {
        use super::*;
        #[derive(Template, Deserialize, Serialize)]
        #[template(path = "manifests/example.yaml.jinja")]
        pub struct Example {
            pub resources: Vec<Resource>,
        }
    }
}

// Operator Templates
pub mod operator {
    use super::*;

    #[derive(Template)]
    #[template(path = "operator/main.rs.jinja")]
    pub struct Main {
        pub api_group: String,
        pub api_version: String,
        pub controllers: Vec<String>,
        pub types: Vec<String>,
    }

    #[derive(Template)]
    #[template(path = "operator/cli.rs.jinja")]
    pub struct Cli {
        pub project_name: String,
        pub version: String,
        pub author: String,
    }

    #[derive(Template)]
    #[template(path = "operator/controller.rs.jinja")]
    pub struct Controller {
        pub tag: String,
        pub arg_name: String,
        pub kind_struct: String,
        pub dto_fields: Vec<Field>,
        pub resource_remote_ref: String,
        pub has_create_action: bool,
        pub has_update_action: bool,
        pub has_delete_action: bool,
        pub api_url: String,
    }

    #[derive(Template)]
    #[template(path = "operator/controller_action_delete.jinja")]
    pub struct ControllerActionDelete<'a> {
        pub arg_name: String,
        pub kind_struct: String,
        pub controllers: Vec<&'a ControllerAttributes>,
        pub resource_remote_ref: String,
    }

    #[derive(Template)]
    #[template(path = "operator/controller_action_update.jinja")]
    pub struct ControllerActionPut<'a> {
        pub arg_name: String,
        pub kind_struct: String,
        pub controllers: Vec<&'a ControllerAttributes>,
        pub resource_remote_ref: String,
    }

    #[derive(Template)]
    #[template(path = "operator/controller_action_create.jinja")]
    pub struct ControllerActionPost<'a> {
        pub arg_name: String,
        pub kind_struct: String,
        pub controllers: Vec<&'a ControllerAttributes>,
        pub resource_remote_ref: String,
    }

    #[derive(Template)]
    #[template(path = "operator/type.rs.jinja")]
    pub struct Type {
        pub tag_name: String,
        pub type_name: String,
        pub api_version: String,
        pub group_name: String,
        pub fields: Vec<Field>,
        pub reference_id: String,
    }

    #[derive(Template)]
    #[template(path = "operator/lib.rs.jinja")]
    pub struct Lib {}
}

// Cargo Templates
pub mod cargo {
    use super::*;

    #[derive(Template)]
    #[template(path = "cargo.toml.jinja")]
    pub struct CargoToml {}

    #[derive(Template)]
    #[template(path = "operator/cargo.toml.jinja")]
    pub struct OperatorCargoToml {}

    #[derive(Template)]
    #[template(path = "crdgen/cargo.toml.jinja")]
    pub struct CrdgenCargoToml {}

    #[derive(Template)]
    #[template(path = "tests/cargo.toml.jinja")]
    pub struct TestsCargoToml {}
}

// Crdgen Templates
pub mod crdgen {
    use super::*;

    #[derive(Template)]
    #[template(path = "crdgen/main.rs.jinja")]
    pub struct Main {
        pub resources: BTreeMap<String, String>,
    }
}

// General Templates
pub mod general {
    use super::*;

    #[derive(Template)]
    #[template(path = ".dockerignore.jinja")]
    pub struct Dockerignore {}

    #[derive(Template)]
    #[template(path = ".editorconfig.jinja")]
    pub struct Editorconfig {}

    #[derive(Template)]
    #[template(path = ".env.example.jinja")]
    pub struct EnvExample {
        pub operator_name: String,
        pub operator_author: String,
        pub operator_api_group: String,
        pub operator_api_version: String,
        pub operator_resource_ref: String,
        pub operator_example_metadata_spec_field_ref: String,
        pub operator_include_tags: String,
        pub operator_secret_name: String,
    }

    #[derive(Template)]
    #[template(path = ".gitattributes.jinja")]
    pub struct GitAttributes {}

    #[derive(Template)]
    #[template(path = ".gitignore.jinja")]
    pub struct GitIgnore {}

    #[derive(Template)]
    #[template(path = ".openapi-generator-ignore.jinja")]
    pub struct OpenAPIGeneratorIgnore {}

    #[derive(Template)]
    #[template(path = ".prettierrc.yaml.jinja")]
    pub struct Prettierrc {}

    #[derive(Template)]
    #[template(path = ".rustfmt.toml.jinja")]
    pub struct RustfmtToml {}

    #[derive(Template)]
    #[template(path = "cluster.yaml.jinja")]
    pub struct ClusterYaml {}

    #[derive(Template)]
    #[template(path = "dockerfile.jinja")]
    pub struct Dockerfile {}

    #[derive(Template)]
    #[template(path = "readme.md.jinja")]
    pub struct ReadmeMd {
        pub project_name: String,
    }

    #[derive(Template)]
    #[template(path = "taskfile.yml.jinja")]
    pub struct Taskfile {}

    #[derive(Template)]
    #[template(path = ".cargo/config.toml.jinja")]
    pub struct CargoConfig {}
}
