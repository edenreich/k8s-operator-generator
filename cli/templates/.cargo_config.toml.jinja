# This file is generated by kopgen. Do not edit manually. If you need to make adjustments add it to .openapi-generator-ignore file.
[target.aarch64-unknown-linux-musl]
linker = "aarch64-linux-musl-gcc"
rustflags = ["-C", "target-feature=+crt-static"]

[target.x86_64-unknown-linux-musl]
linker = "x86_64-linux-musl-gcc"
rustflags = ["-C", "target-feature=+crt-static"]
