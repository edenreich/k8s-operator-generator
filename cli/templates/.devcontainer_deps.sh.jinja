#!/usr/bin/zsh

set -e

sudo apt-get update && sudo apt-get install -y pkg-config libssl-dev

# Install kopgen CLI
curl -sSL https://github.com/edenreich/k8s-operator-generator/releases/download/v1.0.1/kopgen_$(uname -m)-unknown-linux-musl -o kopgen
chmod +x kopgen
sudo mv kopgen /usr/local/bin

# Install NVM and NodeJS
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.0/install.sh | bash
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
nvm install --lts --latest-npm
npm install -g prettier

# Install favorite prompt theme
git clone --depth=1 https://github.com/romkatv/powerlevel10k.git ${ZSH_CUSTOM:-$HOME/.oh-my-zsh/custom}/themes/powerlevel10k

# Install task
TASK_VERSION=v3.39.2
curl -s https://taskfile.dev/install.sh | sudo sh -s -- -b /usr/local/bin $TASK_VERSION

# Install ctlptl
CTLPTL_VERSION="0.8.34"
curl -fsSL https://github.com/tilt-dev/ctlptl/releases/download/v$CTLPTL_VERSION/ctlptl.$CTLPTL_VERSION.linux.x86_64.tar.gz | sudo tar -xzv -C /usr/local/bin ctlptl

# Install k3d
K3D_VERSION="v5.7.4"
curl -s https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | TAG=$K3D_VERSION bash

# Install kubectl
KUBECTL_VERSION="v1.31.0"
curl -L https://dl.k8s.io/release/$KUBECTL_VERSION/bin/linux/amd64/kubectl -o ~/kubectl
chmod +x ~/kubectl
sudo mv ~/kubectl /usr/local/bin/kubectl

# Install mdbook
cargo install mdbook
