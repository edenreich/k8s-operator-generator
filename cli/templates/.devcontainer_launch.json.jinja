{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "lldb",
      "request": "launch",
      "name": "Debug executable 'operator'",
      "env": {
        "RUST_BACKTRACE": "1",
        "RUST_LOG": "debug",
        "ACCESS_TOKEN": ""
      },
      "cargo": {
        "args": ["build", "--package=k8s_operator"]
      },
      "args": [],
      "cwd": "${workspaceFolder}"
    },
    {
      "type": "lldb",
      "request": "launch",
      "name": "Debug kopgen: init command",
      "env": {
        "RUST_BACKTRACE": "1",
        "RUST_LOG": "debug",
        "ACCESS_TOKEN": ""
      },
      "cargo": {
        "args": ["build", "--package=kopgen"]
      },
      "args": ["init"],
      "cwd": "${workspaceFolder}"
    },
    {
      "type": "lldb",
      "request": "launch",
      "name": "Debug codegen: hydrate command",
      "env": {
        "RUST_BACKTRACE": "1",
        "RUST_LOG": "debug",
        "ACCESS_TOKEN": ""
      },
      "cargo": {
        "args": ["build", "--package=kopgen"]
      },
      "args": ["hydrate", "openapi.yaml"],
      "cwd": "${workspaceFolder}"
    },
    {
      "type": "lldb",
      "request": "launch",
      "name": "Debug codegen: generate command",
      "env": {
        "RUST_BACKTRACE": "1",
        "RUST_LOG": "debug",
        "ACCESS_TOKEN": ""
      },
      "cargo": {
        "args": ["build", "--package=kopgen"]
      },
      "args": ["generate", "openapi.yaml"],
      "cwd": "${workspaceFolder}"
    }
  ]
}
