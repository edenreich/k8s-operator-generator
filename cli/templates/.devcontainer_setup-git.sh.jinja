#!/bin/bash

set -e

git config --global --add safe.directory /workspaces/kopgen
