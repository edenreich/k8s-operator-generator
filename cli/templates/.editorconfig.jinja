# This file is generated by kopgen. Do not edit manually. If you need to make adjustments add it to .openapi-generator-ignore file.
root = true

[*]
charset = utf-8
end_of_line = lf
indent_size = 4
indent_style = space
insert_final_newline = true
trim_trailing_whitespace = true

[*.{yaml,yml,json}]
indent_style = space
indent_size = 2

[*.md]
trim_trailing_whitespace = false
indent_size = 2

[*.rs]
indent_style = space
indent_size = 4
