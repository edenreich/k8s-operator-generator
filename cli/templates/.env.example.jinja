KUBERNETES_OPERATOR_GROUP=example.com
KUBERNETES_OPERATOR_VERSION=v1
KUBERNETES_OPERATOR_RESOURCE_REF=uuid
KUBERNETES_OPERATOR_EXAMPLE_METADATA_SPEC_FIELD_REF=name
KUBERNETES_OPERATOR_INCLUDE_TAGS=cats

OPENAPI_DOWNLOAD_URL=https://raw.githubusercontent.com/edenreich/kopgen/refs/heads/main/openapi.yaml
INSTALL_CRDS=true
RUST_LOG=info
RUST_BACKTRACE=1
CONTAINER_REGISTRY=localhost:5005
CLUSTER_NAME=k3d-k3s-default
TARGET_ARCH=x86_64-unknown-linux-musl
RELEASE=false
