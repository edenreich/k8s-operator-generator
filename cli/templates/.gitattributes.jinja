# This file is generated by kopgen. Do not edit manually. If you need to make adjustments add it to .openapi-generator-ignore file.
crates/client-sdk/** linguist-generated=true
crates/k8s-crdgen/** linguist-generated=true
crates/k8s-operator/** linguist-generated=true
docs/openapi/** linguist-generated=true
manifests/** linguist-generated=true
tests/** linguist-generated=true

.devcontainer/** linguist-vendored=true
Dockerfile linguist-vendored=true
docs/** linguist-vendored=true
