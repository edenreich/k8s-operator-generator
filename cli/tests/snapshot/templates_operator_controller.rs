use askama::Template;
use insta::assert_snapshot;
use kopgen::templates::{operator::Controller, Field};

#[test]
fn render() {
    let template = Controller {
        tag: "example_tag".to_string(),
        arg_name: "argName".to_string(),
        kind_struct: "ExampleKind".to_string(),
        dto_fields: vec![
            Field {
                pub_name: "field1".to_string(),
                field_type: "string".to_string(),
            },
            Field {
                pub_name: "field2".to_string(),
                field_type: "string".to_string(),
            },
        ],
        resource_remote_ref: "resourceRef".to_string(),
        has_create_action: true,
        has_update_action: true,
        has_delete_action: false,
        api_url: "https://api.example.com".to_string(),
    };
    let rendered = template.render().expect("Failed to render template");
    assert_snapshot!(rendered);
}
