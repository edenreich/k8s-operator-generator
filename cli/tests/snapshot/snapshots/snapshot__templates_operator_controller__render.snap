---
source: cli/tests/snapshot/templates_operator_controller.rs
expression: rendered
---
// This file is generated by kopgen. Do not edit manually. If you need to make adjustments add it to .openapi-generator-ignore file.
use std::{sync::Arc, time::Duration};

use futures::StreamExt;
use kube::api::{Api, PostParams, Resource};
use kube_runtime::{controller::Action, watcher, Controller};
use log::{error, info, warn};

use openapi::{
    apis::{
        example_tag_api::{
            create_argName,
            delete_argName_by_id,
            get_argName_by_id,
            update_argName_by_id
        },
        configuration::Configuration,
    },
    models::ExampleKind as ExampleKindDto,
};

use crate::{
    errors::OperatorError,
    types::argName::{
        ExampleKind,
        ExampleKindSpec,
        ExampleKindStatus,
    },
    {add_finalizer, create_condition, remove_finalizer, update_status},
};

const REQUEUE_AFTER_IN_SEC: u64 = 30;
const API_URL: &str = "https://api.example.com";
const API_USER_AGENT: &str = "k8s-operator";

struct ExtraArgs {
    kube_client: Api<ExampleKind>,
}

pub async fn handle(kube_client: Api<ExampleKind>) -> Result<(), OperatorError> {
    info!("Starting the controller");
    let controller = Controller::new(kube_client.clone(), watcher::Config::default());

    let extra_args = Arc::new(ExtraArgs {
        kube_client: kube_client.clone(),
    });

    info!("Running the controller");
    controller
        .run(reconcile, error_policy, extra_args)
        .for_each(|res| async {
            match res {
                Ok(action) => info!("Reconciliation was successful, action: {:?}", action),
                Err(e) => error!("Error reconciling: {:?}", e),
            }
        })
        .await;

    info!("ExampleKind Controller has stopped");
    Ok(())
}

async fn reconcile(argName: Arc<ExampleKind>, ctx: Arc<ExtraArgs>) -> Result<Action, OperatorError> {
    let kube_client = ctx.kube_client.clone();
    let mut argName = argName.as_ref().clone();
    let resourceRef = argName
        .status
        .as_ref()
        .and_then(|status| status.resourceRef.clone())
        .unwrap_or_default();

    if argName.status.is_none() {
        add_default_status(&kube_client, &mut argName).await?;
    }

    if argName.meta().deletion_timestamp.is_some() {
        handle_delete(&kube_client, &mut argName, &resourceRef).await?;
    } else if resourceRef.is_empty() {
        handle_create(&kube_client, &mut argName).await?;
    } else if argName.meta().generation != argName.status.as_ref().unwrap().observed_generation {
        handle_update(&kube_client, &mut argName, &resourceRef).await?;
    }

    check_for_drift(&kube_client, &mut argName).await?;
    Ok(Action::requeue(Duration::from_secs(REQUEUE_AFTER_IN_SEC)))
}

async fn get_client_config() -> Result<Configuration, OperatorError> {
    let config = Configuration {
        base_path: API_URL.to_string(),
        client: reqwest::Client::new(),
        user_agent: Some(API_USER_AGENT.to_string()),
        bearer_access_token: Some(std::env::var("ACCESS_TOKEN").unwrap_or_default()),
        ..Default::default()
    };
    Ok(config)
}

async fn add_default_status(kube_client: &Api<ExampleKind>, argName: &mut ExampleKind) -> Result<(), OperatorError> {
    argName.status = Some(ExampleKindStatus {
        conditions: vec![],
        resourceRef: None,
        observed_generation: Some(0),
    });
    update_status(kube_client, argName.clone())
        .await
        .map_err(|e| OperatorError::FailedToUpdateStatus(e.into()))
}

pub async fn check_for_drift(
    kube_client: &Api<ExampleKind>,
    argName: &mut ExampleKind,
) -> Result<(), OperatorError> {
    let dto = converters::kube_type_to_dto(argName.clone());
    let resourceRef = converters::resourceRef_to_string(dto.resourceRef).unwrap_or_default();
    let config = get_client_config().await?;

    if resourceRef.is_empty() {
        warn!("ExampleKind has no status, cannot get by id or check for drift. Skipping...");
        return Ok(());
    }

    match get_argName_by_id(&config, &resourceRef).await {
        Ok(dto) => {
            let remote_argName = converters::dto_to_kube_type(dto);
            if remote_argName != argName.spec {
                let current_argName_dto = converters::kube_type_to_dto(argName.clone());
                warn!("ExampleKind has drifted remotely, sending an update to remote...");
                match update_argName_by_id(&config, &resourceRef, current_argName_dto).await {
                    Ok(_) => {
                        info!("ExampleKind updated successfully");
                        let condition = create_condition(
                            "Updated",
                            "AvailableUpdated",
                            "Updated the resource",
                            "Resource has been updated",
                            argName.meta().generation,
                        );
                        let mut argName_clone = argName.clone();
                        if let Some(status) = argName_clone.status.as_mut() {
                            status.conditions.push(condition);
                            status.observed_generation = argName.meta().generation;
                        }
                        update_status(kube_client, argName_clone).await?
                    }
                    Err(e) => {
                        error!("Failed to update ExampleKind: {:?}", e);
                        return Err(OperatorError::FailedToGetResource(e.into()));
                    }
                }
            }
        }
        Err(e) => {
            error!("Failed to get ExampleKind: {:?}", e);
            return Err(OperatorError::FailedToGetResource(e.into()));
        }
    }

    Ok(())
}

fn error_policy(_resource: Arc<ExampleKind>, error: &OperatorError, _ctx: Arc<ExtraArgs>) -> Action {
    error!("Error processing event: {:?}", error);
    Action::requeue(Duration::from_secs(REQUEUE_AFTER_IN_SEC))
}

mod converters {
    use super::{
        ExampleKind, ExampleKindDto, ExampleKindSpec
    };

    pub fn resourceRef_to_string(resourceRef: Option<resourceRef::Uuid>) -> Option<String> {
        resourceRef.map(|resourceRef| resourceRef.to_string())
    }

    fn string_to_resourceRef(resourceRef: Option<String>) -> Option<resourceRef::Uuid> {
        match resourceRef {
            Some(resourceRef) => match resourceRef::Uuid::parse_str(&resourceRef) {
                Ok(resourceRef) => Some(resourceRef),
                Err(_) => None,
            },
            None => None,
        }
    }

    pub fn kube_type_to_dto(argName: ExampleKind) -> ExampleKindDto {
        let resourceRef = match argName.status {
            Some(status) => string_to_resourceRef(status.resourceRef),
            None => None,
        };
        ExampleKindDto {
            resourceRef,
                field1: argName.spec.field1,
                field2: argName.spec.field2,
        }
    }

    pub fn dto_to_kube_type(argName: ExampleKindDto) -> ExampleKindSpec {
        ExampleKindSpec {
            field1: argName.field1,
            field2: argName.field2,
        }
    }
}
